src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -230,71 +231,36 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
-
-            while (current != null)
-            {
-                Debug.Assert(current.IsUnconvertedInterpolatedStringAddition);
-                stack.Push(current);
-                isConstant = isConstant && current.Right.ConstantValue is not null;
-                var rightInterpolatedString = (BoundUnconvertedInterpolatedString)current.Right;
-
-                if (!ValidateInterpolatedStringParts(rightInterpolatedString))
+            if (!binaryOperator.VisitBinaryOperatorInterpolatedString(partsArrayBuilder, static (BoundUnconvertedInterpolatedString unconvertedInterpolatedString, ArrayBuilder<ImmutableArray<BoundExpression>> partsArrayBuilder) =>
                 {
-                    // Exception to case 3. Delegate to standard binding.
-                    stack.Free();
-                    partsArrayBuilder.Free();
-                    return false;
-                }
-
-                partsCount += rightInterpolatedString.Parts.Length;
-                partsArrayBuilder.Add(rightInterpolatedString.Parts);
-
-                switch (current.Left)
-                {
-                    case BoundBinaryOperator leftOperator:
-                        current = leftOperator;
-                        continue;
-                    case BoundUnconvertedInterpolatedString interpolatedString:
-                        isConstant = isConstant && interpolatedString.ConstantValue is not null;
-
-                        if (!ValidateInterpolatedStringParts(interpolatedString))
-                        {
-                            // Exception to case 3. Delegate to standard binding.
-                            stack.Free();
-                            partsArrayBuilder.Free();
-                            return false;
-                        }
+                    if (!ValidateInterpolatedStringParts(unconvertedInterpolatedString))
+                    {
+                        return false;
+                    }
 
-                        partsCount += interpolatedString.Parts.Length;
-                        partsArrayBuilder.Add(interpolatedString.Parts);
-                        current = null;
-                        break;
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
-                }
+                    partsArrayBuilder.Add(unconvertedInterpolatedString.Parts);
+                    return true;
+                }))
+            {
+                partsArrayBuilder.Free();
+                return false;
             }
 
-            Debug.Assert(partsArrayBuilder.Count == stack.Count + 1);
             Debug.Assert(partsArrayBuilder.Count >= 2);
 
             if (isConstant ||
-                (partsCount <= 4 && partsArrayBuilder.All(static parts => AllInterpolatedStringPartsAreStrings(parts))))
+                (partsArrayBuilder.Count <= 4 && partsArrayBuilder.All(static parts => AllInterpolatedStringPartsAreStrings(parts))))
```

Comment:
>AllInterpolatedStringPartsAreStrings [](http://example.com/codeflow?start=87&length=36)

Does creating a lambda rather than just referencing `AllInterpolatedStringPartsAreStrings` avoid delegate type instantiations here?

Reply:
Yes, we never cache method references.